segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/OperationProcessor.java

```diff
@@ -392,66 +401,213 @@ private boolean isFatalException(Throwable ex) {
+                            toFail.put(op, ex);
+
+                            // Then fail the remaining operations (which also handles fatal errors) and bail out.
+                            collectFailureCandidates(ex, commitArgs, toFail);
+                            if (isFatalException(ex)) {
+                                this.stopException = ex;
+                                CallbackHelpers.invokeSafely(this.fatalExceptionCallback, ex, null);
+                            }
+
+                            return;
+                        }
+
+                        if (this.stopException != null) {
+                            // Even if the operation succeeded, we encountered a stop exception (fatal failure) and most likely
+                            // this operation will be corrupted in the DurableDataLog. Do not ack it as success. If the owning
+                            // Container does end up recovering successfully, it's up to the client to figure out the state
+                            // of the affected segments so it can resume operations.
+                            toFail.put(op, this.stopException);
+                        } else {
+                            toComplete.add(op);
+                        }
+                    }
+
+                    this.highestCommittedDataFrame = addressSequence;
                 }
 
-                e.complete();
+                this.logUpdater.flush();
+            } finally {
+                toComplete.forEach(CompletableOperation::complete);
+                toFail.forEach(this::failOperation);
+                autoCompleteIfNeeded();
+                if (toFail.size() == 0) {
+                    // Only record the commit if we had no failures.
+                    this.checkpointPolicy.recordCommit(commitArgs.getDataFrameLength());
+                }
             }
-
-            this.logUpdater.flush();
-            this.checkpointPolicy.recordCommit(commitArgs.getDataFrameLength());
         }
 
         /**
          * Rolls back all pending Metadata changes and fails all pending operations.
          *
          * @param ex The cause of the failure. The operations will be failed with this as a cause.
+         * @param commitArgs The Data Frame Commit Args that triggered this action.
+         */
+        void fail(Throwable ex, DataFrameBuilder.CommitArgs commitArgs) {
+            Map<CompletableOperation, Throwable> toFail = new HashMap<>();
+            try {
+                synchronized (this.lock) {
+                    collectFailureCandidates(ex, commitArgs, toFail);
+                }
+            } finally {
+                toFail.forEach(this::failOperation);
+                autoCompleteIfNeeded();
+            }
+
+            if (isFatalException(ex)) {
+                synchronized (this.lock) {
+                    this.stopException = ex;
```

Comment:
`collectFailureCandidates` is immediately followed by a statement that fails all of those operations, regardless of whether the exception was fatal or not (we need to fail them anyway). The line of code you point to is simply an inspector that determines if a special callback (into the OperationProcessor) should be invoked to notify of that fact.